src/internal/document_api/mod.rs

```diff
@@ -902,216 +907,197 @@ fn document_create<'a>(
+    )?;
+
+    aes::decrypt(&mut enc_doc, *sym_key.bytes())
+        .map_err(|e| e.into())
+        .map(move |decrypted_doc| DocumentDecryptResult {
+            id: doc_meta.0.id,
+            name: doc_meta.0.name,
+            created: doc_meta.0.created,
+            updated: doc_meta.0.updated,
+            decrypted_data: decrypted_doc.to_vec(),
         })
 }
 
 /// Decrypt the unmanaged document. The caller must provide both the encrypted data as well as the
 /// encrypted DEKs. Most use cases would want `decrypt_document` instead.
-pub fn decrypt_document_unmanaged<'a, CR: rand::CryptoRng + rand::RngCore>(
-    auth: &'a RequestAuth,
-    recrypt: &'a Recrypt<Sha256, Ed25519, RandomBytes<CR>>,
-    device_private_key: &'a PrivateKey,
-    encrypted_doc: &'a [u8],
-    encrypted_deks: &'a [u8],
-) -> impl Future<Item = DocumentDecryptUnmanagedResult, Error = IronOxideErr> + 'a {
+pub async fn decrypt_document_unmanaged<CR: rand::CryptoRng + rand::RngCore>(
+    auth: &RequestAuth,
+    recrypt: &Recrypt<Sha256, Ed25519, RandomBytes<CR>>,
+    device_private_key: &PrivateKey,
+    encrypted_doc: &[u8],
+    encrypted_deks: &[u8],
+) -> Result<DocumentDecryptUnmanagedResult, IronOxideErr> {
     // attempt to parse the proto as fail-fast validation. If it fails decrypt will fail
-    protobuf::parse_from_bytes::<EncryptedDeksP>(encrypted_deks)
-        .map_err(IronOxideErr::from)
-        .into_future()
-        .join(parse_document_parts(encrypted_doc))
-        .and_then(
-            |(
-                proto_edeks,
-                (
-                    DocumentHeader {
-                        document_id,
-                        segment_id,
-                    },
-                    aes_encrypted_value,
-                ),
-            )| {
-                if document_id.id() == proto_edeks.get_documentId()
-                    && segment_id as i32 == proto_edeks.get_segmentId()
-                {
-                    Ok((document_id, aes_encrypted_value))
-                } else {
-                    Err(IronOxideErr::UnmanagedDecryptionError(
-                        proto_edeks.get_documentId().into(),
-                        proto_edeks.get_segmentId(),
-                        document_id.0,
-                        segment_id as i32,
-                    ))
-                }
-            },
-        )
-        .join(requests::edek_transform::edek_transform(
-            &auth,
-            encrypted_deks,
+
+    let ((proto_edeks, (doc_meta, mut aes_encrypted_value)), transform_resp) = try_join!(
+        async {
+            Ok((
+                protobuf::parse_from_bytes::<EncryptedDeksP>(encrypted_deks)
+                    .map_err(IronOxideErr::from)?,
+                parse_document_parts(encrypted_doc)?,
+            ))
+        },
+        requests::edek_transform::edek_transform(&auth, encrypted_deks,)
+    )?;
+
+    let _ = edeks_and_header_match_or_err(&proto_edeks, &doc_meta)?;
```

Comment:
The assignment isn't necessary as the Result has been dealt with by the `?`.
```suggestion
    edeks_and_header_match_or_err(&proto_edeks, &doc_meta)?;
```